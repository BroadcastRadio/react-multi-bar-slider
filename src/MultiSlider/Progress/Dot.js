import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import transition from '../utils/transition';

const StyledDot = glamorous.span({
  position: 'absolute',
  display: 'block',
  zIndex: 5,
  borderRadius: '50%'
}, ({ hasIcon, width, height, color, noTransition }) => ({
  top: hasIcon ? 0 : '50%',
  transform: hasIcon ? 'translateX(-50%)' : 'translate(50%, -50%)',
  width: hasIcon ? 0 : width,
  height: hasIcon ? 0 : height,
  backgroundColor: hasIcon ? 'transparent' : color,
  transition: noTransition ? 'none' : transition
}));

const DotIcon = glamorous.img({
  position: 'absolute',
  transform: 'translateX(-50%)',
  userDrag: 'none',
  userSelect: 'none'
}, ({ width, height }) => ({
  width,
  height
}));

const Dot = ({ dot, sliderColor, mouseDown }) => (
  <StyledDot
    className="dot"
    hasIcon={!!dot.icon}
    width={dot.width || 28}
    height={dot.height || 28}
    color={dot.color || sliderColor}
    css={dot.style}
    noTransition={mouseDown}
  >
    {dot.icon && (
      <DotIcon
        className="icon"
        src={dot.icon}
        width={dot.width || 50}
        height={dot.height || 50}
        draggable={false}
        css={dot.iconStyle}
      />
    )}
  </StyledDot>
);

Dot.propTypes = {
  dot: PropTypes.oneOfType([
    PropTypes.shape({
      width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      color: PropTypes.string,
      icon: PropTypes.string,
      style: PropTypes.object
    }),
    PropTypes.bool
  ]).isRequired,
  sliderColor: PropTypes.string.isRequired,
  mouseDown: PropTypes.bool.isRequired
};

export default Dot;
