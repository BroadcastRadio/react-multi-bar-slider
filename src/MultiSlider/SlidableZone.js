import PropTypes from 'prop-types';
import styled from '@emotion/styled';

const SlidableZone = styled('div')({
  position: 'absolute',
  width: '100%'
}, ({ size, zIndex }) => ({
  top: -size,
  bottom: -size,
  zIndex
}));

SlidableZone.propTypes = {
  size: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  zIndex: PropTypes.number.isRequired
};

export default SlidableZone;
