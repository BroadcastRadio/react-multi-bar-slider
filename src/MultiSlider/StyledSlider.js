import PropTypes from 'prop-types';
import styled from 'react-emotion';
import transition from '../utils/transition';
import getHalf from '../utils/getHalf';

const StyledSlider = styled('div')({
  position: 'relative',
  height: 14,
  boxSizing: 'border-box',
  transition
}, ({ width, height, backgroundColor, roundedCorners, readOnly, css }) => ({
  width,
  height,
  backgroundColor,
  borderRadius: roundedCorners ? getHalf(height) : 0,
  cursor: readOnly ? 'auto' : 'pointer',
  ...css
}));


StyledSlider.propTypes = {
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  backgroundColor: PropTypes.string.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  readOnly: PropTypes.bool.isRequired,
  css: PropTypes.object.isRequired
};

StyledSlider.defaultProps = {
  css: null
};

export default StyledSlider;
