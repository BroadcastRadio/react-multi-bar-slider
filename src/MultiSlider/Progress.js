import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import sliderPropType from './utils/sliderPropType';
import glamorous from 'glamorous';
import transition from './utils/transition';
import getHalf from './utils/getHalf';
import Dot from './Dot';

const StyledProgress = glamorous.div(
  {
    position: 'absolute',
    top: 0,
    right: 0
  },
  ({
    color,
    progress,
    height,
    equal,
    equalColor,
    roundedCorners,
    zIndex,
    noTransition
  }) => ({
    width: `${progress || 0}%`,
    height,
    backgroundColor: equal && equalColor ? equalColor : color,
    borderRadius: roundedCorners ? getHalf(height) : 0,
    zIndex,
    transition: noTransition ? 'none' : transition
  })
);

const Progress = ({
  slider,
  height,
  slidersEqual,
  equalColor,
  roundedCorners,
  mouseDown,
  zIndex
}) => (
  <Fragment>
    <StyledProgress
      className="progress"
      color={slider.color}
      progress={slider.progress}
      height={height}
      equal={slidersEqual}
      equalColor={equalColor}
      roundedCorners={roundedCorners}
      zIndex={zIndex}
      noTransition={mouseDown}
    />
    {slider.dot && (
      <Dot
        dot={slider.dot}
        progress={slider.progress}
        sliderColor={slider.color}
        mouseDown={mouseDown}
      />
    )}
  </Fragment>
);

Progress.propTypes = {
  slider: sliderPropType,
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  slidersEqual: PropTypes.bool.isRequired,
  equalColor: PropTypes.string.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  mouseDown: PropTypes.bool.isRequired,
  zIndex: PropTypes.number.isRequired
};

export default Progress;
