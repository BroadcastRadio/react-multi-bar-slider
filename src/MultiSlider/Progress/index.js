import React from 'react';
import PropTypes from 'prop-types';
import sliderPropType from '../utils/sliderPropType';
import Dot from './Dot';
import processStyle from '../utils/processStyle';
import StyledProgress from './StyledProgress';

const Progress = ({
  slider,
  height,
  slidersEqual,
  equalColor,
  roundedCorners,
  mouseDown,
  zIndex
}) => (
  <StyledProgress
    className="progress"
    color={slider.color}
    progress={slider.progress}
    height={height}
    equal={slidersEqual}
    equalColor={equalColor}
    roundedCorners={roundedCorners}
    noTransition={mouseDown}
    zIndex={zIndex}
    css={processStyle(slider.style, {
      slider,
      height,
      slidersEqual,
      equalColor,
      roundedCorners,
      mouseDown,
      zIndex
    })}
  >
    {slider.dot && (
      <Dot
        dot={slider.dot}
        sliderColor={slider.color}
        mouseDown={mouseDown}
      />
    )}
  </StyledProgress>
);

Progress.propTypes = {
  slider: sliderPropType,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  slidersEqual: PropTypes.bool.isRequired,
  equalColor: PropTypes.string,
  roundedCorners: PropTypes.bool.isRequired,
  mouseDown: PropTypes.bool.isRequired,
  zIndex: PropTypes.number.isRequired
};

export default Progress;
