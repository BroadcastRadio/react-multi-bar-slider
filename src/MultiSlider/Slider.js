import React from 'react';
import PropTypes from 'prop-types';
import processStyle from '../utils/processStyle';
import StyledSlider from './StyledSlider';

const Slider = ({
  width,
  height,
  backgroundColor,
  style,
  onMouseMoveActivate,
  onMouseMoveDeactivate,
  onMouseMove,
  roundedCorners,
  readOnly,
  children,
  ...props
}) => (
  <StyledSlider
    width={width}
    height={height}
    backgroundColor={backgroundColor}
    roundedCorners={roundedCorners}
    readOnly={readOnly}
    css={processStyle(style, {
      width,
      height,
      backgroundColor,
      roundedCorners,
      readOnly
    })}
    onMouseDown={onMouseMoveActivate}
    onMouseUp={onMouseMoveDeactivate}
    onMouseLeave={onMouseMoveDeactivate}
    onMouseMove={onMouseMove}
    {...props}
  >
    {children}
  </StyledSlider>
);

Slider.displayName = 'Slider';

Slider.propTypes = {
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  backgroundColor: PropTypes.string.isRequired,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  onMouseMoveActivate: PropTypes.func.isRequired,
  onMouseMoveDeactivate: PropTypes.func.isRequired,
  onMouseMove: PropTypes.func.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  readOnly: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired
};

export default Slider;
