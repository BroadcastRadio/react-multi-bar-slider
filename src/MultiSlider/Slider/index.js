import React from 'react';
import PropTypes from 'prop-types';
import processStyle from '../utils/processStyle';
import StyledSlider from './StyledSlider';

const Slider = ({
  width,
  height,
  backgroundColor,
  style,
  onSlide,
  onMouseMoveActivate,
  onMouseMoveDeactivate,
  onMouseMove,
  roundedCorners,
  readOnly,
  children
}) => (
  <StyledSlider
    width={width}
    height={height}
    backgroundColor={backgroundColor}
    roundedCorners={roundedCorners}
    readOnly={readOnly}
    css={processStyle(style, {
      width,
      height,
      backgroundColor,
      roundedCorners,
      readOnly
    })}
    onClick={onSlide}
    onMouseDown={onMouseMoveActivate}
    onMouseUp={onMouseMoveDeactivate}
    onMouseLeave={onMouseMoveDeactivate}
    onMouseMove={onMouseMove}
  >
    {children}
  </StyledSlider>
);

Slider.propTypes = {
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  backgroundColor: PropTypes.string.isRequired,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  onSlide: PropTypes.func.isRequired,
  onMouseMoveActivate: PropTypes.func.isRequired,
  onMouseMoveDeactivate: PropTypes.func.isRequired,
  onMouseMove: PropTypes.func.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  readOnly: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired
};

export default Slider;
