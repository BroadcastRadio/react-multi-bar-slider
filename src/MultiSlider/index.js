import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import getProgressFromMousePosition from './utils/getProgressFromMousePosition';
import slidersEqual from './utils/slidersEqual';
import sortSliders from './utils/sortSliders';
import sliderPropType from './utils/sliderPropType';
import Slider from './Slider';
import Progress from './Progress';
import SlidableZone from './SlidableZone';

export default class MultiSlider extends PureComponent {
  static propTypes = {
    sliders: PropTypes.arrayOf(sliderPropType).isRequired,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    slidableZoneSize: PropTypes.number,
    backgroundColor: PropTypes.string,
    equalColor: PropTypes.string,
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    onSlide: PropTypes.func.isRequired,
    roundedCorners: PropTypes.bool,
    reverse: PropTypes.bool,
    readOnly: PropTypes.bool
  };

  static defaultProps = {
    width: '100%',
    height: 14,
    slidableZoneSize: 7,
    backgroundColor: '#EEEEEE',
    roundedCorners: false,
    style: {},
    reversed: false,
    readOnly: false
  };

  state = { mouseDown: false };

  handleSlide = (e) => {
    const { onSlide, reversed, readOnly } = this.props;
    if (readOnly) return;

    const newProgress = getProgressFromMousePosition(e, reversed);
    onSlide(newProgress);
  };

  handleMouseMoveActivate = () => this.setState({ mouseDown: true });
  handleMouseMoveDeactivate = () => this.setState({ mouseDown: false });

  handleMouseMove = (e) => {
    if (!this.state.mouseDown) return;
    this.handleSlide(e);
  };

  render = () => {
    const {
      sliders,
      width,
      height,
      slidableZoneSize,
      backgroundColor,
      equalColor,
      style,
      roundedCorners,
      reversed,
      readOnly
    } = this.props;
    const allSlidersEqual = slidersEqual(sliders);
    const sortedSliders = sortSliders(sliders);

    return (
      <Slider
        width={width}
        height={height}
        backgroundColor={backgroundColor}
        style={style}
        onSlide={this.handleSlide}
        onMouseMoveActivate={this.handleMouseMoveActivate}
        onMouseMoveDeactivate={this.handleMouseMoveDeactivate}
        onMouseMove={this.handleMouseMove}
        roundedCorners={roundedCorners}
        readOnly={readOnly}
      >
        {sliders.map((slider, i) =>
          <Progress
            slider={slider}
            height={height}
            slidersEqual={allSlidersEqual}
            equalColor={equalColor}
            roundedCorners={roundedCorners}
            reversed={reversed}
            mouseDown={this.state.mouseDown}
            zIndex={sortedSliders.indexOf(slider)}
            key={i}
          />
        )}
        <SlidableZone size={slidableZoneSize} zIndex={sliders.length} />
      </Slider>
    );
  };
}
