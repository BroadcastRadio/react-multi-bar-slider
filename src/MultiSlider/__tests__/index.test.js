import React from 'react';
import Enzyme, { shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import MultiSlider from '../index';
import Slider from '../Slider';
import Progress from '../Progress';
import SlidableZone from '../SlidableZone';

Enzyme.configure({ adapter: new Adapter() });

beforeAll(() => {
  global.multiSliderProps = {
    sliders: [
      {
        color: '#00BDAF',
        progress: 17,
        dot: {
          icon: 'foo.png',
          iconStyle: { bottom: -7 }
        }
      },
      {
        color: '#AB47BC',
        progress: 45,
        style: ({ slidersEqual }) => ({
          borderTopLeftRadius: slidersEqual ? 7 : 0
        }),
        dot: {
          icon: 'bar.png',
          iconStyle: { top: 7 }
        }
      }
    ],
    slidableZoneSize: 50,
    onSlide: () => {}
  };
});

afterAll(() => {
  delete global.multiSliderProps;
});

describe('MultiSlider.js', () => {
  it('matches the snapshot', () => {
    const tree = renderer
      .create(<MultiSlider {...multiSliderProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const multiSlider = render(<MultiSlider {...multiSliderProps} />);
    expect(multiSlider).toBeDefined();
  });

  it('renders a Slider', () => {
    const multiSlider = shallow(<MultiSlider {...multiSliderProps} />);
    expect(multiSlider.find(Slider).length).toBe(1);
  });

  it('renders a Progress for every given slider', () => {
    const multiSlider = shallow(<MultiSlider {...multiSliderProps} />);
    expect(multiSlider.find(Progress).length).toBe(2);
  });

  it('renders a SlidableZone', () => {
    const multiSlider = shallow(<MultiSlider {...multiSliderProps} />);
    expect(multiSlider.find(SlidableZone).length).toBe(1);
  });

  it('handles slider clicks correctly', () => {
    const onSlide = jest.fn();
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps} onSlide={onSlide} />
    );
    multiSlider.find(Slider).prop('onSlide')({
      target: {
        classList: [],
        getBoundingClientRect: () => ({
          left: 154,
          width: 876
        })
      },
      pageX: 933.64
    });
    expect(onSlide.mock.calls[0][0]).toBe(11);
  });

  it('handles mouse moves correctly', () => {
    const onSlide = jest.fn();
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps} onSlide={onSlide} />
    );
    const fakeEvent = {
      target: {
        classList: [],
        getBoundingClientRect: () => ({
          left: 154,
          width: 876
        })
      },
      pageX: 933.64
    };
    multiSlider.find(Slider).prop('onMouseMove')(fakeEvent);
    expect(onSlide.mock.calls.length).toBe(0);

    multiSlider.find(Slider).prop('onMouseMoveActivate')();
    multiSlider.update();
    multiSlider.find(Slider).prop('onMouseMove')(fakeEvent);
    expect(onSlide.mock.calls[0][0]).toBe(11);

    multiSlider.find(Slider).prop('onMouseMoveDeactivate')();
    multiSlider.update();
    multiSlider.find(Slider).prop('onMouseMove')(fakeEvent);
    expect(onSlide.mock.calls.length).toBe(1);
  });
});
