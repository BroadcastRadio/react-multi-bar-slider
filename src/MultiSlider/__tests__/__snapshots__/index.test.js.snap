// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSlider.js matches the snapshot 1`] = `
<div
  className="css-1i6y0pe"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  readOnly={false}
>
  <div
    className="css-18897st progress"
  >
    <span
      className="css-14o13ui dot"
    >
      <img
        className="css-va7boi icon"
        draggable={false}
        src="foo.png"
      />
    </span>
  </div>
  <div
    className="css-ay0g0g progress"
  >
    <span
      className="css-14o13ui dot"
    >
      <img
        className="css-q037tf icon"
        draggable={false}
        src="bar.png"
      />
    </span>
  </div>
  <div
    className="css-qfp31e"
  />
</div>
`;
